Client/Frontend/Browser/Tabs/RemoteTabs/RemoteTabsPanelMiddleware.swift

```diff
@@ -0,0 +1,71 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import UIKit
+import Storage
+import Common
+import Shared
+import Redux
+
+class RemoteTabsPanelMiddleware {
+    private let profile: Profile
+
+    init(profile: Profile = AppContainer.shared.resolve()) {
+        self.profile = profile
+    }
+
+    lazy var remoteTabsPanelProvider: Middleware<AppState> = { state, action in
+        switch action {
+        case RemoteTabsPanelAction.refreshTabs:
+            self.refreshTabs(updateCache: true)
+        case RemoteTabsPanelAction.refreshCachedTabs:
+            self.refreshTabs(updateCache: false)
+        default:
+            break
+        }
+    }
+
+    // MARK: - Internal Utilities
+
+    private func refreshTabs(updateCache: Bool = false) {
+        ensureMainThread { [self] in
+            guard profile.hasSyncableAccount() else {
+                DispatchQueue.main.async {
```

Comment:
@OrlaM @yoanarios I was seeing a bug with the Redux updates on the Sync tab in some flows, and I realized that what was happening is that basically the order of the state updates arrives in an unexpected way if you call `store.dispatch()` from the middleware while already in the midst of handling a previous `store.dispatch()` for another action.

So basically it causes issues because (afaics) Redux will resolve the 2nd (newer) action before the first/older parent action is fully resolved. But that means that the newer action's state update eventually gets "overridden" by the previous/older action, and so you wind up with your state updates happening in an unexpected/out-of-order way.

In this PR I temporarily fixed the problem by async-dispatching the 2nd call to `store.dispatch` here. But I think this might be worth discussing further to decide out how best to handle this type of situation (whether we want to e.g. enforce a rule for this so that it isn't possible to do that, or handle it some other way). This _seems_ like a valid use case but I also could be misunderstanding something, but either way I could foresee other folks in the future trying this same thing, so it might be good to have a plan for handling that.

Reply:
I wonder if it would make sense for actions to be added to some type of internal serial queue so that the current action is always fully resolved before any subsequent ones are. (Offhand I can't really think of a situation where you would want an action dispatched from a middleware to be immediately processed before the parent action was finished resolving, which is basically what happens right now AFAIU)

Reply:
That's a conversation that I had with @OrlaM during the implementation but at that point, we decided to go with the pattern of always dispatching the actions from middleware which solved the issue but it's not off the table for more complicated flows to introduce some handling like the one you mention

Reply:
Ok makes sense; thanks @yoanarios for clarifying 👍

Reply:
Ensuring the middleware always dispatches back to the main thread even if it doesn't leave the main thread was a quick way to move forward, with the idea to potentially iterate in the future. I can definitely see this resulting in bugs at some point because a dispatch call gets missed. 
One of the options we threw out was making the store a main actor so that events dispatched to it would always have to be called async, it's possible that could result in other less desirable complications tho. 
It's not urgent to get back to, and the workaround of dispatching to the main thread is fine for now but I would like to see this eventually evolve into something where human error is less easy to introduce. 
There is a spike ticket here to look into the main actor option if you find yourself with some free time feel free to investigate, or try out other ideas if you have some. https://mozilla-hub.atlassian.net/browse/FXIOS-7093

Reply:
Thanks @OrlaM 👍 Sounds good, will take a look